cmd/frontend/graphqlbackend/external_services_test.go

```diff
@@ -281,6 +282,256 @@ func TestAddExternalService(t *testing.T) {
+		marshaledUserID := MarshalUserID(userID)
+
+		t.Run("service kind is not supported", func(t *testing.T) {
+			result, err := newSchemaResolver(db).AddExternalService(ctx, &addExternalServiceArgs{
+				Input: addExternalServiceInput{
+					Namespace:   &marshaledUserID,
+					Kind:        "BITBUCKET",
+					DisplayName: "Bitbucket",
+				},
+			})
+
+			if want := backend.ErrExternalServiceKindNotSupported; err != want {
+				t.Errorf("got err %v, want %v", err, want)
+			}
+			if result != nil {
+				t.Errorf("got result %v, want nil", result)
+			}
+		})
+
+		t.Run("user can still add external services", func(t *testing.T) {
+			svcs := []*types.ExternalService{
+				{Kind: extsvc.KindGitHub, NamespaceOrgID: orgID},
+			}
+
+			externalServices := database.NewMockExternalServiceStore()
+			externalServices.ListFunc.SetDefaultReturn(svcs, nil)
+			db.ExternalServicesFunc.SetDefaultReturn(externalServices)
+
+			_, err := newSchemaResolver(db).AddExternalService(ctx, &addExternalServiceArgs{
+				Input: addExternalServiceInput{
+					Namespace:   &marshaledUserID,
+					Kind:        "GITLAB",
+					DisplayName: "GitLab",
+					Config:      "{\n  \"url\": \"https://gitlab.com\",\n  \"token\": \"dfdf\",\n  \"projectQuery\": [\n    \"projects?membership=true&archived=no\"\n  ]\n}",
+				},
+			})
+			if err != nil {
+				t.Fatal(err)
+			}
+		})
+
+		t.Run("user has reached the limit for a given kind of service", func(t *testing.T) {
+			svcs := []*types.ExternalService{
+				{Kind: extsvc.KindGitLab, NamespaceOrgID: orgID},
+			}
+
+			externalServices := database.NewMockExternalServiceStore()
+			externalServices.ListFunc.SetDefaultReturn(svcs, nil)
+			db.ExternalServicesFunc.SetDefaultReturn(externalServices)
+
+			result, err := newSchemaResolver(db).AddExternalService(ctx, &addExternalServiceArgs{
+				Input: addExternalServiceInput{
+					Namespace:   &marshaledUserID,
+					Kind:        "GITLAB",
+					DisplayName: "GitLab",
+					Config:      "{\n  \"url\": \"https://gitlab.com\",\n  \"token\": \"dfdf\",\n  \"projectQuery\": [\n    \"projects?membership=true&archived=no\"\n  ]\n}",
+				},
+			})
+			if want := backend.ErrExternalServiceLimitPerKindReached; err != want {
+				t.Errorf("got err %v, want %v", err, want)
+			}
+			if result != nil {
+				t.Errorf("got result %v, want nil", result)
+			}
+		})
+
+		t.Run("user has reached the limit for both types of service", func(t *testing.T) {
+			svcs := []*types.ExternalService{
+				{Kind: extsvc.KindGitLab, NamespaceOrgID: orgID},
+				{Kind: extsvc.KindGitHub, NamespaceOrgID: orgID},
+			}
+
+			externalServices := database.NewMockExternalServiceStore()
+			externalServices.ListFunc.SetDefaultReturn(svcs, nil)
+			db.ExternalServicesFunc.SetDefaultReturn(externalServices)
+
+			result, err := newSchemaResolver(db).AddExternalService(ctx, &addExternalServiceArgs{
+				Input: addExternalServiceInput{
+					Namespace:   &marshaledUserID,
+					Kind:        "GITLAB",
```

Comment:
I think we have a constant for this:

```suggestion
					Kind:        extsvc.KindGitLab,
```

Same for other occurrences.